File: src/android/LineLogin.java
package plugin.line;

import com.linecorp.linesdk.LineAccessToken;
import com.linecorp.linesdk.LineApiResponse;
import com.linecorp.linesdk.LineApiResponseCode;
import com.linecorp.linesdk.LineIdToken;
import com.linecorp.linesdk.LineProfile;
import com.linecorp.linesdk.Scope;
import com.linecorp.linesdk.api.LineApiClient;
import com.linecorp.linesdk.api.LineApiClientBuilder;
import com.linecorp.linesdk.auth.LineAuthenticationParams;
import com.linecorp.linesdk.auth.LineLoginApi;
import com.linecorp.linesdk.auth.LineLoginResult;

import android.content.Intent;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Arrays;

public class LineLogin extends CordovaPlugin {

    private static final int REQUEST_CODE = 1;
    private static final int PARAMETER_ERROR = -1;
    private static final int SDK_ERROR = -2;
    private static final int UNKNOWN_ERROR = -3;

    private String channelId;
    private CallbackContext loginCallbackContext;
    private static LineApiClient lineApiClient;

    @Override
    public boolean execute(String action, JSONArray data, CallbackContext callbackContext) throws JSONException {

        if (action.equals("initialize")) {
            this.initialize(data, callbackContext);
            return true;
        } else if (action.equals("login")) {
            this.login(callbackContext, false);
            return true;
        } else if (action.equals("loginWeb")) {
            this.login(callbackContext, true);
            return true;
        } else if (action.equals("logout")) {
            this.logout(callbackContext);
            return true;
        } else if (action.equals("getAccessToken")) {
            this.getAccessToken(callbackContext);
            return true;
        } else if (action.equals("verifyAccessToken")) {
            this.verifyAccessToken(callbackContext);
            return true;
        } else if (action.equals("refreshAccessToken")) {
            this.refreshAccessToken(callbackContext);
            return true;
        } else {
            return false;
        }
    }

    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode != REQUEST_CODE) {
            return;
        }

        CallbackContext callbackContext = loginCallbackContext;
        loginCallbackContext = null;
        if (callbackContext == null) {
            return;
        }

        LineLoginResult result = LineLoginApi.getLoginResultFromIntent(data);
        if (result.getResponseCode() == LineApiResponseCode.SUCCESS) {
            JSONObject json = new JSONObject();
            try {
                LineProfile profile = result.getLineProfile();
                if (profile != null) {
                    json.put("userID", profile.getUserId());
                    json.put("displayName", profile.getDisplayName());
                }
                if (profile != null && profile.getPictureUrl() != null) {
                    json.put("pictureURL", profile.getPictureUrl().toString());
                }

                LineIdToken lineIdToken = result.getLineIdToken();
                if (lineIdToken != null && lineIdToken.getEmail() != null) {
                    json.put("email", lineIdToken.getEmail());
                }

                callbackContext.success(json);
            } catch (JSONException e) {
                this.unknownError(callbackContext, e.toString());
            }
        } else if (result.getResponseCode() == LineApiResponseCode.CANCEL) {
            this.sdkError(callbackContext, result.getResponseCode().toString(), "user cancel");
        } else {
            this.sdkError(callbackContext, result.getResponseCode().toString(), result.toString());
        }
    }

    private void initialize(JSONArray data, CallbackContext callbackContext) throws JSONException {
        if (data.length() == 0 || data.isNull(0)) {
            this.parameterError(callbackContext, "channel_id is required.");
            return;
        }

        JSONObject params = data.getJSONObject(0);
        channelId = params.optString("channel_id", "");
        if (channelId.length() == 0) {
            this.parameterError(callbackContext, "channel_id is required.");
        } else {
            LineApiClientBuilder apiClientBuilder = new LineApiClientBuilder(this.cordova.getActivity().getApplicationContext(), channelId);
            lineApiClient = apiClientBuilder.build();
            callbackContext.success();
        }
    }

    private void login(CallbackContext callbackContext, boolean webOnly) {
        try {
            if (!this.isInitialized()) {
                this.parameterError(callbackContext, "initialize must be called before login.");
                return;
            }

            LineAuthenticationParams authenticationParams = new LineAuthenticationParams.Builder()
                    .scopes(Arrays.asList(Scope.PROFILE, Scope.OPENID_CONNECT, Scope.OC_EMAIL))
                    .build();
            Intent loginIntent = webOnly
                    ? LineLoginApi.getLoginIntentWithoutLineAppAuth(
                            this.cordova.getActivity().getApplicationContext(),
                            channelId,
                            authenticationParams
                    )
                    : LineLoginApi.getLoginIntent(
                            this.cordova.getActivity().getApplicationContext(),
                            channelId,
                            authenticationParams
                    );
            loginCallbackContext = callbackContext;
            this.cordova.startActivityForResult((CordovaPlugin) this, loginIntent, REQUEST_CODE);
        } catch (Exception e) {
            this.unknownError(callbackContext, e.toString());
        }
    }

    private void logout(CallbackContext callbackContext) {
        this.runLineApiCall(callbackContext, new Runnable() {
            @Override
            public void run() {
                LineLogin.this.lineApiClient.logout();
                callbackContext.success();
            }
        });
    }

    private void getAccessToken(CallbackContext callbackContext) {
        this.runLineApiCall(callbackContext, new Runnable() {
            @Override
            public void run() {
                try {
                    JSONObject json = new JSONObject();
                    LineAccessToken lineAccessToken = lineApiClient.getCurrentAccessToken().getResponseData();
                    if (lineAccessToken == null) {
                        LineLogin.this.sdkError(callbackContext, "NO_ACCESS_TOKEN", "No current LINE access token.");
                        return;
                    }
                    json.put("accessToken", lineAccessToken.getTokenString());
                    json.put("expireTime", lineAccessToken.getEstimatedExpirationTimeMillis());
                    callbackContext.success(json);
                } catch (JSONException e) {
                    LineLogin.this.unknownError(callbackContext, e.toString());
                }
            }
        });
    }

    private void verifyAccessToken(CallbackContext callbackContext) {
        this.runLineApiCall(callbackContext, new Runnable() {
            @Override
            public void run() {
                LineApiResponse verifyResponse = lineApiClient.verifyToken();
                if (verifyResponse.isSuccess()) {
                    callbackContext.success();
                } else {
                    LineLogin.this.sdkError(
                            callbackContext,
                            verifyResponse.getResponseCode().toString(),
                            verifyResponse.getErrorData() == null ? "" : verifyResponse.getErrorData().toString()
                    );
                }
            }
        });
    }

    private void refreshAccessToken(CallbackContext callbackContext) {
        this.runLineApiCall(callbackContext, new Runnable() {
            @Override
            public void run() {
                LineAccessToken lineAccessToken = lineApiClient.refreshAccessToken().getResponseData();
                if (lineAccessToken == null) {
                    LineLogin.this.sdkError(callbackContext, "NO_ACCESS_TOKEN", "LINE access token could not be refreshed.");
                    return;
                }
                callbackContext.success(lineAccessToken.getTokenString());
            }
        });
    }

    private boolean isInitialized() {
        return channelId != null && channelId.length() > 0 && lineApiClient != null;
    }

    private void runLineApiCall(final CallbackContext callbackContext, final Runnable runnable) {
        if (!this.isInitialized()) {
            this.parameterError(callbackContext, "initialize must be called before calling LINE APIs.");
            return;
        }

        this.cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    runnable.run();
                } catch (Exception e) {
                    LineLogin.this.unknownError(callbackContext, e.toString());
                }
            }
        });
    }

    private void parameterError(CallbackContext callbackContext, String description) {
        try {
            JSONObject json = new JSONObject();
            json.put("code", PARAMETER_ERROR);
            json.put("description", description);
            callbackContext.error(json);
        } catch (JSONException e) {
            this.unknownError(callbackContext, e.toString());
        }
    }

    private void sdkError(CallbackContext callbackContext, String sdkErrorCode, String description) {
        try {
            JSONObject json = new JSONObject();
            json.put("code", SDK_ERROR);
            json.put("sdkErrorCode", sdkErrorCode);
            json.put("description", description);
            callbackContext.error(json);
        } catch (JSONException e) {
            this.unknownError(callbackContext, e.toString());
        }
    }

    private void unknownError(CallbackContext callbackContext, String description) {
        try {
            JSONObject json = new JSONObject();
            json.put("code", UNKNOWN_ERROR);
            json.put("description", description);
            callbackContext.error(json);
        } catch (JSONException e) {
            callbackContext.error(-1);
        }
    }
}

File: src/ios/LineLogin.swift
import Foundation
import LineSDK

@objc(LineLogin) class Line : CDVPlugin {
    
    struct ErrorCode {
        static let ParameterError = -1
        static let SDKError = -2
        static let UnknownError = -3
    }

    @objc func initialize(_ command: CDVInvokedUrlCommand) {
        
        guard command.arguments.count > 0, let params = command.arguments[0] as? [String: Any] else {
            self.parameterError(command: command, description: "channel_id is required")
            return
        }
        
        guard let channelID = params["channel_id"] as? String, !channelID.isEmpty else {
            self.parameterError(command: command, description: "channel_id is required")
            return
        }
        
        if !LoginManager.shared.isSetupFinished {
            LoginManager.shared.setup(channelID: channelID, universalLinkURL: nil)
        }
        let result = CDVPluginResult(status: CDVCommandStatus_OK)
        commandDelegate.send(result, callbackId:command.callbackId)
    }
    
    func _login(_ command: CDVInvokedUrlCommand, onlyWebLogin: Bool) {
        guard LoginManager.shared.isSetupFinished else {
            self.parameterError(command: command, description: "initialize must be called before login")
            return
        }

        var parameters = LoginManager.Parameters()
        parameters.onlyWebLogin = onlyWebLogin
        LoginManager.shared.login(permissions: [.profile, .openID, .email], in: self.viewController, parameters: parameters) {
            result in
            switch result {
            case .success(let loginResult):
                var data = [String : Any]()
                
                if let displayName = loginResult.userProfile?.displayName {
                    data["displayName"] = displayName
                }
                if let userID = loginResult.userProfile?.userID {
                    data["userID"] = userID
                }
                if let email = loginResult.accessToken.IDToken?.payload.email {
                    data["email"] = email
                }
                if let pictureURL = loginResult.userProfile?.pictureURL {
                    data["pictureURL"] = String(describing: pictureURL)
                }
                
                let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:data)
                self.commandDelegate.send(result, callbackId:command.callbackId)
            case .failure(let error):
                self.sdkError(command: command, error: error)
            }
        }
    }
    
    @objc func loginWeb(_ command: CDVInvokedUrlCommand) {
        self._login(command, onlyWebLogin: true)
    }
    
    @objc func login(_ command: CDVInvokedUrlCommand) {
        self._login(command, onlyWebLogin: false)
    }
    
    @objc func logout(_ command: CDVInvokedUrlCommand) {
        LoginManager.shared.logout { result in
            switch result {
            case .success:
                let result = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate.send(result, callbackId:command.callbackId)
            case .failure(let error):
                self.sdkError(command: command, error: error)
            }
        }
    }

    @objc func getAccessToken(_ command: CDVInvokedUrlCommand) {
        
        guard let currentAccessToken = AccessTokenStore.shared.current else {
            self.sdkError(command: command, sdkErrorCode: "NO_ACCESS_TOKEN", description: "No current LINE access token.")
            return
        }
        
        let data = ["accessToken":currentAccessToken.value, "expireTime":currentAccessToken.expiresAt.timeIntervalSince1970] as [String : Any]
        let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:data)
        commandDelegate.send(result, callbackId:command.callbackId)
    }
    
    @objc func verifyAccessToken(_ command: CDVInvokedUrlCommand) {
        
        API.Auth.verifyAccessToken { (result) in
            switch result {
            case .success( _):
                let result = CDVPluginResult(status: CDVCommandStatus_OK)
                self.commandDelegate.send(result, callbackId:command.callbackId)
            case .failure(let error):
                self.sdkError(command: command, error: error)
            }
        }
    }
    
    @objc func refreshAccessToken(_ command: CDVInvokedUrlCommand) {

        API.Auth.refreshAccessToken { (result) in
            switch result {
            case .success(let accessToken):
                let data = ["accessToken":accessToken.value, "expireTime":accessToken.expiresAt.timeIntervalSince1970] as [String : Any]
                let result = CDVPluginResult(status: CDVCommandStatus_OK, messageAs:data)
                self.commandDelegate.send(result, callbackId:command.callbackId)
            case .failure(let error):
                self.sdkError(command: command, error: error)
            }
        }
    }
    
    private func parameterError(command: CDVInvokedUrlCommand, description: String) {
        let err = ["code":ErrorCode.ParameterError, "description": description] as [AnyHashable : Any]
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err)
        self.commandDelegate.send(result, callbackId:command.callbackId)
    }
    
    private func sdkError(command: CDVInvokedUrlCommand, error: LineSDKError) {
        let description = error.errorDescription ?? ""
        self.sdkError(command: command, sdkErrorCode: String(error.errorCode), description: description)
    }

    private func sdkError(command: CDVInvokedUrlCommand, sdkErrorCode: String, description: String) {
        let err = ["code":ErrorCode.SDKError, "sdkErrorCode":sdkErrorCode, "description": description] as [AnyHashable : Any]
        let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: err)
        self.commandDelegate.send(result, callbackId:command.callbackId)
    }
    
}
